[ContactStructuralMechanicsApplication] Adding Octree with OBB search (used by default)

loumalouomega opened the pull request:
This is a huge PR, and depends on #4263 (which depends on #4264, #4221 , #4222 and #4262). Feel free to block this until the others are merged

The idea is to have a better initial estimation of the potential pairs, right now obtained with a radius search, in order to improve the conditioning of the system and robustness of the method

I still have some problems with the initial octree search, it gives some problem when it is used only with conditions (that's why I needed to update some tests in order to make it work). I am thinking on combine the radius search with the OBB, so I will have many initial pairs with a proper filtering from OBB

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Change name

loumalouomega committed:
Correct

loumalouomega committed:
TDim description

loumalouomega committed:
Mark method as const

loumalouomega committed:
Changing name of the class

loumalouomega committed:
Changing name class

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
2D objects are not supposed to have faces

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Update header

loumalouomega committed:
[skip ci] Minor clean up

loumalouomega committed:
Making it virtual

loumalouomega committed:
FindIntersectedGeometricalObjectsWithOBBForSearchProcess WIP

loumalouomega committed:
Making more modular the base class

loumalouomega committed:
Making public

loumalouomega committed:
Advances

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Correction of typos

loumalouomega committed:
Minor

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding constructor requested by @pooyan-dadvand

loumalouomega committed:
Adding SeparatingAxisTheoremHasIntersection

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[ContactStructuralMechanicsApplication] Tests relative paths

Fixes a relative path

loumalouomega committed:
[StructuralMechanicsApplication] Commenting debug GidIO generation

This comments a post binary created for debug reasons

loumalouomega committed:
Separating the Octree and KDTree search into separate functions

loumalouomega committed:
Creating auxiliar method

loumalouomega committed:
Fixing stuff

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
[ContactStructuralMechanicsApplication] Fixing linking problem

This fixes after #4399. (In Windows it worked :P, apparently in Linux gives a linking problem on execution)

loumalouomega committed:
[ContactStructuralMechanicsApplication] Typo on python scripts for search

search_search -> search_utility_list

loumalouomega committed:
Better

loumalouomega committed:
Transforming search utilities into process

loumalouomega committed:
MIssing overrides

loumalouomega committed:
Adding wrapper for search

loumalouomega committed:
Removing dependency of new methods

loumalouomega committed:
More clean up

loumalouomega committed:
Removing new methods

philbucher committed:
Pythonic (I)

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega committed:
Corrections

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
Minor test clean up

loumalouomega committed:
Removing unused

loumalouomega committed:
Clean up tests

loumalouomega committed:
More clean up

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Minor change on search

loumalouomega committed:
Minor

loumalouomega committed:
CPP test for search (WIP)

loumalouomega committed:
Trigger appveyor

loumalouomega committed:
Minor

loumalouomega committed:
Clean up

loumalouomega committed:
Some changes

loumalouomega committed:
Moving to protected variables

loumalouomega committed:
Remove unused parameters

loumalouomega committed:
Update

loumalouomega committed:
Update test

loumalouomega committed:
Correcting test

loumalouomega committed:
Clean unused parameters

loumalouomega committed:
Missing const

loumalouomega committed:
Missing Info, PrintInfo, PrintData

loumalouomega committed:
Avoid comparison compilation error in Travis

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding check

loumalouomega committed:
Setting NODAL_H manually

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Improving debug info

loumalouomega committed:
Adding intersection type to the process

loumalouomega committed:
Updating processes

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[Core] Minor clean up of MortarUtilities (also reducing debug verbosity)

This is just a fast clean up of the MortarUtilities, also reducing debug verbosity

loumalouomega committed:
Avoiding repeated conditions

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
More debug info

loumalouomega committed:
Adding more complex case

loumalouomega committed:
Flag  setting fix

loumalouomega committed:
Fixing test

loumalouomega committed:
Updating Parameters

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Extending BB

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Using Octree search by default on MeshTying

loumalouomega committed:
Covering corner cases

loumalouomega committed:
ALmost all tests working with OBB

loumalouomega committed:
Clean up

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
Clean up

loumalouomega committed:
Some clean up

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Improving HasIntersection from OctreeBinary

loumalouomega committed:
Missing argument

loumalouomega committed:
More clean up

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Moving to protected

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding lower_bounding_box_coefficient and  higher_bounding_box_coefficient as parameters

loumalouomega committed:
Last tests working now with Octree

AlejandroCornejo commented:
how is it going @loumalouomega ? Now it works more or less?

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,665 @@
 //                     Kratos default license: kratos/license.txt
 //
-//  Main authors:    Pooyan dadvand
+//  Main authors:    Vicente Mataix Ferrandiz
```
I don't think you are the main and unique author of this...

loumalouomega replied:
This is probably for bad copy paste, changing. In any case this is in #4263

loumalouomega replied:
Indeed, in the cpp the authorship is preserved at it is, this is result of some copy paste

~~~cpp
//  Main authors:    Pooyan Davand
//  Collaborators:   Ruben Zorrilla Martinez
//                   Vicente Mataix Ferrandiz
//
~~~

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,665 @@
+     * @tparam TEntity The type of geometrical entity considered (if conditions or elements)
+     * @author Pooyan Dadvand
+     */
+    template<class TEntity = Element>
```
Why is this template argument now needed? It is spread over all the derived classes that use this process.

loumalouomega replied:
Because right now it only works with elements, I did this change already in #4263

rubenzorrilla replied:
I know, this always requires to define the skin as an element based mesh. The point is that I don't know if this was done intentionally.

loumalouomega replied:
In any case if you check, the python exposition remains the same. It only affects from the cpp side

rubenzorrilla reviewed (changes requested):
I think that this PR can be split in multiple minor PRs. This will simplify quite a lot the review process, specially taking into account that there is a lot of aesthetic changes.

loumalouomega commented:
> I think that this PR can be split in multiple minor PRs. This will simplify quite a lot the review process, specially taking into account that there is a lot of aesthetic changes.

It is already splitted in multiple PR, read the first comment please

rubenzorrilla commented:
So which is the purpose of this one?

loumalouomega committed:
Revert authorship

loumalouomega commented:
> So which is the purpose of this one?

It is explained in the first comment, adding this search to the ContactStructuralMechanicsApplication

rubenzorrilla commented:
So then why don't we wait to the other PRs to be in the master branch to create this one?

loumalouomega commented:
> So then why don't we wait to the other PRs to be in the master branch to create this one?

Because I just finished with this one, the other ones are taking time, I need to focus in other projects. I prefer to open as a reminder

rubenzorrilla commented:
As you prefer, but this means that any changed applied during the review process of the other PRs would need to be added to this PR. This makes easy to left any of the changes proposed from now on, specially taking into account that there are several "subPRs" involved.

loumalouomega commented:
> So then why don't we wait to the other PRs to be in the master branch to create this one?

Besides, this is quite common:

Note that #2919 it is required for #2933 and #4169. The review of these PR are focus in the parts not affected by #2919

loumalouomega commented:
> As you prefer, but this means that any changed applied during the review process of the other PRs would need to be added to this PR. This makes easy to left any of the changes proposed from now on, specially taking into account that there are several "subPRs" involved.

I know, I am was already updating the other PR. As I said this is what happened in #2919 for #2933 and #4169.

loumalouomega commented:
> how is it going @loumalouomega ? Now it works more or less?

It works, just I am not 100% satisfied at this point. I needed to update some tests (checking that the result was OK). And I wasn't happy with that

loumalouomega committed:
Extending list and typo

loumalouomega committed:
Adding geometry constructors

loumalouomega committed:
Using constructor to reduce code duplication

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Minor correction

loumalouomega committed:
Adding KdTree with OBB intersecion

loumalouomega committed:
Adding test

loumalouomega committed:
in_radius_with_obb as default search method

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
Correction

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Missing KRATOS_API

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Missing template argument

loumalouomega committed:
Fix

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Moving FastCrossProduct to MathUtils

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega commented on kratos/spatial_containers/octree_binary.h:
```diff
@@ -1171,7 +1171,7 @@ namespace Kratos {
         while (!(delta_z & (one << min_level)) && (min_level > min_level_2)) min_level--;
         min_level++;
-
+        
```
```suggestion

```

loumalouomega committed:
Reverting whitespace

loumalouomega committed:
Solving conflict

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Deleting assigment operator

loumalouomega committed:
Default destructor

loumalouomega committed:
Moving Parameters to base class

loumalouomega committed:
Moving to this class the parameters

loumalouomega committed:
Removing Parameters

loumalouomega committed:
Reducing code duplication

loumalouomega committed:
More deterministic test

loumalouomega committed:
More deterministic test for contact

loumalouomega committed:
Update files concerning ContactStructuralMechanicsApplication

loumalouomega committed:
Updating process for OBB

loumalouomega committed:
Update name of model parts

loumalouomega committed:
More code reduction

loumalouomega committed:
Updating name on Contact application

loumalouomega committed:
Revert "Updating name on Contact application"

This reverts commit 1101a5b27444a9d63de493b8e2fa543ee4509bcf.

loumalouomega committed:
Revert "Updating name for OBB"

This reverts commit c57ac33d8e7d200f038f27e8e0df00f378bd1527.

loumalouomega committed:
Updating name for OBB

loumalouomega committed:
Revert "Updating name for base class"

This reverts commit 9829ec978be6b3f4677242af613a3249ab2f6260.

loumalouomega committed:
Updating name for base class

loumalouomega committed:
Correct name

loumalouomega committed:
Fix

loumalouomega committed:
Windows compilation

loumalouomega committed:
Linux compilation issue

loumalouomega committed:
Adding template argument to octree_binary

loumalouomega committed:
Generalizing combinations

@rubenzorrilla requested

loumalouomega committed:
Adding test

loumalouomega committed:
Extending to OBB process

loumalouomega committed:
Expose to python (I)

loumalouomega committed:
Expose to python (II)

loumalouomega committed:
Reverting python exposition

loumalouomega committed:
Reverting math_utils

loumalouomega committed:
Using CrossProduct

rubenzorrilla committed:
Delete

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

rubenzorrilla committed:
Delete

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding entities utilities

loumalouomega committed:
Reducing  explicit templates

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Revert spaces changes

loumalouomega committed:
Reverting specifications

loumalouomega committed:
Reverting changes

loumalouomega committed:
Update test

loumalouomega committed:
Refactoring without templates (I)

loumalouomega committed:
Refactoring without templates (II)

loumalouomega committed:
Refactoring for contact

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Missing override

loumalouomega committed:
Reverting changes geometries

loumalouomega committed:
Adding direct intersection (I)

loumalouomega committed:
Adding direct intersection (II)

loumalouomega committed:
Updating search

rubenzorrilla committed:
Missing const

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega committed:
@rubenzorrilla comment

loumalouomega committed:
Extending to contact app

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
[Core] Moving flag from Condition/Element to GeometricalObject

Basically that. The flag is common between the geometries and conditions, but currently both derive from Flag instaed of bein included on the GeometricalObject

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Revert "Extending to contact app"

This reverts commit 482cf0c58b80ae4b4cc4f11fecfb6dec5d0764b4.

loumalouomega committed:
Revert "Updating search"

This reverts commit c893ce60548d0a7aee17bce1623c6050e56139fb.

loumalouomega committed:
Revert "Refactoring for contact"

This reverts commit ff014578088db0c675378d318db27e4875832034.

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adapting to contact

loumalouomega committed:
Reducing overhead of Element/Condition

loumalouomega committed:
Serializer correction

loumalouomega committed:
Exposing Flags to python

loumalouomega committed:
@jcotela commentaries + doxygen

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding local flags

loumalouomega committed:
Adding flags

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Adding flags to contact

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega commented:
Now waiting for #4263

loumalouomega committed:
Reordering flags (start on 0)

loumalouomega committed:
Using flags

loumalouomega committed:
Updating test

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Removing comment

loumalouomega committed:
Updating constructor

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega committed:
Lower extend

loumalouomega committed:
Merge branch 'master' into contact/update-search

loumalouomega commented:
@AlejandroCornejo and @rubenzorrilla this PR now only affects the ContactStructuralMechanicsApplication

rubenzorrilla commented on applications/ContactStructuralMechanicsApplication/custom_processes/find_intersected_geometrical_objects_with_obb_for_search_process.h:
```diff
@@ -0,0 +1,251 @@
+ * @details Fills the serach set. Only works for Conditions
+ * @author Vicente Mataix Ferrandiz
+*/
+class KRATOS_API(CONTACT_STRUCTURAL_MECHANICS_APPLICATION) FindIntersectedGeometricalObjectsWithOBBForSearchProcess
```
I'd call it `FindIntersectedGeometricalObjectsWithOBBContactSearchProcess` instead of  `FindIntersectedGeometricalObjectsWithOBBForSearchProcess`

loumalouomega replied:
OK

loumalouomega replied:
Done @rubenzorrilla

rubenzorrilla reviewed (dismissed):
Removing my blockage.

loumalouomega committed:
Fix conflict

loumalouomega commented:
I need reapproval, @AlejandroCornejo

loumalouomega committed:
Replace FindIntersectedGeometricalObjectsWithOBBForSearchProcess for FindIntersectedGeometricalObjectsWithOBBContactSearchProcess

loumalouomega commented:
@rubenzorrilla I applied your suggestion

loumalouomega commented:
Finally

pooyan-dadvand commented:
Is it faster than the AABB?

loumalouomega commented:
My comparison was with the radius search, and the result fits better than the radius, that was the important point

pooyan-dadvand commented:
I see your point but I am just curious about a time comparison in contact search if you have some

loumalouomega commented:
I did it, but I haven saved any result. I can redo it again if I have some time

pooyan-dadvand commented:
Please!
